File: 2021/Day3/Elixir/binary_diagnostics.ex
# iex()> Day3.first

defmodule BinaryDiagnostics do
  def load_file() do
    File.read!("../input.txt")
    |> String.split("\n", trim: true)
  end

  def most_common(list) do
    gb = Enum.group_by(list, & &1)
    max = Enum.map(gb, fn {_, val} -> length(val) end) |> Enum.max()
    [most_common] = for {key, val} <- gb, length(val) == max, do: key
    most_common
  end

  def least_common(list) do
    gb = Enum.group_by(list, & &1)
    min = Enum.map(gb, fn {_, val} -> length(val) end) |> Enum.min()
    [least_common] = for {key, val} <- gb, length(val) == min, do: key
    least_common
  end

  def get_result(binary_list, comparator) do
    binary_list
    |> Enum.map(fn bin -> String.split(bin, "", trim: true) end)
    |> Enum.zip_reduce([], fn elements, acc ->
      [comparator.(elements) | acc]
    end)
    |> Enum.reverse()
    |> Enum.join("")
    |> Integer.parse(2)
    |> elem(0)
  end

  @spec first() :: Integer
  def first() do
    binaries = load_file
    gamma = get_result(binaries, &most_common/1)
    epsilon = get_result(binaries, &least_common/1)

    gamma * epsilon
  end

  @spec second() :: Integer
  def second() do
    :second
  end
end

File: 2021/Day3/Python/binary_diagnostics.py
def most_common(lst):
    return max(set(lst), key=lst.count)

def least_common(lst):
    return min(set(lst), key=lst.count)

# Part 1
with open("../input.txt") as f:
    lines = list(map(lambda l: l.strip(), f.readlines()))
    binaries = [[] for _ in list(lines[0])]

    for line in lines:
        for i, bit in enumerate(line):
            binaries[i].append(bit)

    gamma_bin = ""
    epsilon_bin = ""

    for column in binaries:
        print(column)
        gamma_bin += most_common(column)
        epsilon_bin += least_common(column)


    gamma = int(gamma_bin, 2)
    epsilon = int(epsilon_bin, 2)

    print(f"Gamma: {gamma}")
    print(f"Epsilon: {epsilon}")
    print(f"Gamma * Epsilon: {gamma * epsilon}")
    

# Part 2
with open("../input.txt") as f:
    lines = map(lambda l: l.strip(), f.readlines())